### ISSUE TITLE
Ambient: services are unable to communicate (signing gRPC error)

### ISSUE THREAD

**Issue by kovaxur** [istio/istio#52057]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Hi,

I try to move from sidecar mode to ambient mode, everything was working fine with sidecar mode, but with ambient mode, services are unable to communicate. The source and destination services are on different nodes, on the source node, the ztunnel pod shows the following error:

`access    connection complete    src.addr=10.201.47.192:58360 src.workload="XXXX-XXXX-56dd6b5597-r6gn6" src.namespace="test2" src.identity="spiffe://cluster.local/ns/test2/sa/XXXX" dst.addr=10.201.12.26:15008 dst.hbone_addr=10.201.12.26:8080 dst.workload="XXXX-569f798c57-gch22" dst.namespace="test2" dst.identity="spiffe://cluster.local/ns/test2/sa/default" direction="outbound" bytes_sent=0 bytes_recv=0 duration="0ms" error="identity error: signing gRPC error (The request does not have valid authentication credentials): request impersonation authentication failure`

I haven't added any authorization policy or request authentication CR.

There are no network security rules/firewalls between cluster nodes. I couldn't find anything related in the documentation for this error.

### Version

```Text
istioctl version
client version: 1.22.1
control plane version: 1.22.2
data plane version: 1.22.2 (47 proxies)

kubectl version
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4-eks-036c24b
```

### Additional Information

Kubernetes: EKS cluster, v1.29

**Comment by kovaxur** [comment]
On the istiod pod, I see the following errors:

`2024-07-15T13:49:26.974247Z    warn    serverca    impersonation failed for identity spiffe://cluster.local/ns/test2/sa/XXX, error: pod istio-system/ztunnel-6xwtm not found client=10.201.15.3:44144`

But the `ztunnel-6xwtm` pod is up and running.

When I execute the `istioctl x ztunnel-config certificates` command, I get only 4 certificates with "Unavailable" status:

```
CERTIFICATE NAME                                             TYPE     STATUS          VALID CERT     SERIAL NUMBER     NOT AFTER     NOT BEFORE
spiffe://cluster.local/ns/test2/sa/XXXX        NA       Unavailable     false          NA                NA            NA
spiffe://cluster.local/ns/test2/sa/default                   NA       Unavailable     false          NA                NA            NA
spiffe://cluster.local/ns/test2/sa/XXXX                 NA       Unavailable     false          NA                NA            NA
spiffe://cluster.local/ns/test2/sa/XXXX     NA       Unavailable     false          NA                NA            NA
```

I think we should have much more certificates normally.

**Comment by kovaxur** [comment]
I think I found the issue. 

I had a feeling, that there's a problem with the root ca, I deleted it and reinstalled the whole istio ambient system, but the root ca configmap (istio-ca-root-cert) was not re-created. In the helm chart config of istiod, I configured the defaults.discoverySelectors configuration to filter out some namespaces:

```
    discoverySelectors:
      - matchLabels:
          kubernetes.io/metadata.name: test2
      - matchLabels:
          kubernetes.io/metadata.name: ...
      - matchLabels:
          kubernetes.io/metadata.name: ...
      - matchLabels:
          kubernetes.io/metadata.name: ...
```

When this is set, the root-ca configmap in the istio-system namespace won't be created (and I guess re-created when moving from sidecar to ambient?). When I removed this configuration, everything started to work.

**Comment by yardenshoham** [comment]
I didn't set `discoverySelectors`. Any idea why I'm getting `identity error: signing gRPC error (The request does not have valid authentication credentials): request impersonation authentication failure`?

Log;
```
2024-08-11T13:36:17.708175Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.workload.Address"
2024-08-11T13:36:17.708232Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.security.Authorization"
2024-08-11T13:36:17.716503Z    warn    xds::client:xds{id=14}    XDS client connection error: gRPC connection error:status: Unauthenticated, message: "authentication failure", retrying i
n 15s
2024-08-11T13:36:32.730828Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.workload.Address"
2024-08-11T13:36:32.730906Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.security.Authorization"
2024-08-11T13:36:32.738831Z    warn    xds::client:xds{id=15}    XDS client connection error: gRPC connection error:status: Unauthenticated, message: "authentication failure", retrying i
n 15s
2024-08-11T13:36:47.749530Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.workload.Address"
2024-08-11T13:36:47.749588Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.security.Authorization"
2024-08-11T13:36:47.755707Z    warn    xds::client:xds{id=16}    XDS client connection error: gRPC connection error:status: Unauthenticated, message: "authentication failure", retrying i
n 15s
2024-08-11T13:37:02.768871Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.workload.Address"
2024-08-11T13:37:02.768915Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.security.Authorization"
2024-08-11T13:37:02.773716Z    warn    xds::client:xds{id=17}    XDS client connection error: gRPC connection error:status: Unauthenticated, message: "authentication failure", retrying i
n 15s
2024-08-11T13:37:17.848593Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.workload.Address"
2024-08-11T13:37:17.848641Z    info    xds::client    sending initial request    resources=0 type_url="type.googleapis.com/istio.security.Authorization"
2024-08-11T13:37:17.868815Z    info    xds::client:xds{id=18}    Stream established
2024-08-11T13:37:17.873601Z    info    xds::client:xds{id=18}    received response    type_url="type.googleapis.com/istio.workload.Address" size=1046 removes=0
2024-08-11T13:37:17.884099Z    info    xds::client:xds{id=18}    received response    type_url="type.googleapis.com/istio.security.Authorization" size=0 removes=0
2024-08-11T13:37:17.884292Z    info    inpod::workloadmanager    handling new stream
2024-08-11T13:37:17.884363Z    info    readiness    Task 'state manager' complete (126.038629801s), still awaiting 1 tasks
2024-08-11T13:37:17.885107Z    info    inpod::statemanager    pod WorkloadUid("c7c2d944-022b-444a-a87e-a2dca7a7d184") received netns, starting proxy
2024-08-11T13:37:17.885313Z    info    proxy::inbound    listener established    address=:15008 component="inbound" transparent=true
2024-08-11T13:37:17.885385Z    info    proxy::inbound_passthrough    listener established    address=:15006 component="inbound plaintext" transparent=true
2024-08-11T13:37:17.885437Z    info    proxy::outbound    listener established    address=:15001 component="outbound" transparent=true
2024-08-11T13:37:17.885801Z    info    inpod::statemanager    pod WorkloadUid("9fbf881a-4fe6-45f3-9fee-5f714aa11d77") received netns, starting proxy
2024-08-11T13:37:17.885984Z    info    proxy::inbound    listener established    address=:15008 component="inbound" transparent=true
2024-08-11T13:37:17.886052Z    info    proxy::inbound_passthrough    listener established    address=:15006 component="inbound plaintext" transparent=true
2024-08-11T13:37:17.886097Z    info    proxy::outbound    listener established    address=:15001 component="outbound" transparent=true
2024-08-11T13:37:17.886294Z    info    inpod::statemanager    pod received snapshot sent
2024-08-11T13:37:17.886309Z    info    readiness    Task 'workload proxy manager' complete (126.040574865s), marking server ready
2024-08-11T13:37:17.904146Z    error    cert_fetcher    unable to prefetch cert for "spiffe://cluster.local/ns/http-test-istio/sa/sleep", skipping, SigningRequest(Status { code: Unauthen
ticated, message: "request impersonation authentication failure", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None })
2024-08-11T13:37:17.954732Z    error    cert_fetcher    unable to prefetch cert for "spiffe://cluster.local/ns/http-test-istio/sa/httpbin", skipping, SigningRequest(Status { code: Unauth
enticated, message: "request impersonation authentication failure", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None })
2024-08-11T13:37:19.416140Z    info    xds::client:xds{id=18}    received response    type_url="type.googleapis.com/istio.workload.Address" size=1046 removes=2
2024-08-11T13:37:19.432029Z    info    xds::client:xds{id=18}    received response    type_url="type.googleapis.com/istio.security.Authorization" size=0 removes=0
2024-08-11T13:37:19.432425Z    error    cert_fetcher    unable to prefetch cert for "spiffe://cluster.local/ns/http-test-istio/sa/httpbin", skipping, SigningRequest(Status { code: Unauth
enticated, message: "request impersonation authentication failure", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None })
2024-08-11T13:37:19.432542Z    error    cert_fetcher    unable to prefetch cert for "spiffe://cluster.local/ns/http-test-istio/sa/sleep", skipping, SigningRequest(Status { code: Unauthen
ticated, message: "request impersonation authentication failure", metadata: MetadataMap { headers: {"content-type": "application/grpc"} }, source: None })
2024-08-11T13:37:19.503467Z    info    xds::client:xds{id=18}    received response    type_url="type.googleapis.com/istio.workload.Address" size=1 removes=0
2024-08-11T13:37:25.993846Z    error    access    connection complete    src.addr=10.42.11.248:36446 src.workload="sleep-7656cf8794-2h2j5" src.namespace="http-test-istio" src.identity="s
piffe://cluster.local/ns/http-test-istio/sa/sleep" dst.addr=10.42.11.247:15008 dst.hbone_addr=10.42.11.247:8080 dst.service="httpbin.http-test-istio.svc.cluster.local" dst.workload="http
bin-86b8ffc5ff-n5tmx" dst.namespace="http-test-istio" dst.identity="spiffe://cluster.local/ns/http-test-istio/sa/httpbin" direction="outbound" bytes_sent=0 bytes_recv=0 duration="0ms" er
ror="identity error: signing gRPC error (The request does not have valid authentication credentials): request impersonation authentication failure"
2024-08-11T13:37:39.763691Z    info    xds::client:xds{id=18}    received response    type_url="type.googleapis.com/istio.workload.Address" size=1 removes=0
2024-08-11T13:40:27.821469Z    error    access    connection complete    src.addr=10.42.11.248:55642 src.workload="sleep-7656cf8794-2h2j5" src.namespace="http-test-istio" src.identity="s
piffe://cluster.local/ns/http-test-istio/sa/sleep" dst.addr=10.42.11.247:15008 dst.hbone_addr=10.42.11.247:8080 dst.service="httpbin.http-test-istio.svc.cluster.local" dst.workload="http
bin-86b8ffc5ff-n5tmx" dst.namespace="http-test-istio" dst.identity="spiffe://cluster.local/ns/http-test-istio/sa/httpbin" direction="outbound" bytes_sent=0 bytes_recv=0 duration="0ms" er
ror="identity error: signing gRPC error (The request does not have valid authentication credentials): request impersonation authentication failure"
```

**Comment by yardenshoham** [comment]
Solved it. I was deploying Istio on the default namespace instead of `istio-system`, I had to modify `istiod`'s values with:
```yaml
istio-istiod:
  profile: ambient
  pilot:
    env:
      CA_TRUSTED_NODE_ACCOUNTS: "default/ztunnel,kube-system/ztunnel"
```

**Comment by bleggett** [comment]
See https://github.com/istio/istio/pull/51311

`Added `.Values.pilot.trustedZtunnelNamespace` to `istiod` chart. Set this if installing ztunnel to a different namespace from `istiod`. Supersedes `.Values.pilot.env.CA_TRUSTED_NODE_ACCOUNTS` (which is still respected if set)`

**Comment by dan-m8t** [comment]
I have the same issue after I set discoverySelectors in meshConfig:

```
{"level":"warn","time":"2024-09-03T11:29:36.564362Z","scope":"serverca","msg":"impersonation failed for identity spiffe://cluster.local/ns/istio-demo/sa/bookinfo-details, error: pod kube-system/ztunnel-gcv2s not found","client":"10.24.2.188:33476"}
```

I also run ztunnel in kube-system (OpenShift), that worked like a charm with `.Values.pilot.trustedZtunnelNamespace` set to `kube-system/ztunnel` but after I configured discoverySelectors it seems not to work anymore :( 

As I am in a test environment for now, how can I re-create the Istio CA if needed? Or any other hint to fix this? I need to use discoverySelectors.

**Comment by dan-m8t** [comment]
I disabled discoverySelectors and it seems to work again, I also deleted the ztunnel pods:

```
istioctl -i kube-system zc certificates ztunnel-fsr7h
CERTIFICATE NAME                                                                     TYPE     STATUS        VALID CERT     SERIAL NUMBER                        NOT AFTER                NOT BEFORE
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-details                      Leaf     Available     true           51238712165e1d61b0bbabcc072ef89f     2024-09-04T12:22:20Z     2024-09-03T12:20:20Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-details                      Root     Available     true           733ce926e70e263a0e03928d910176d7     2034-07-27T09:58:47Z     2024-07-29T09:58:47Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-productpage                  Leaf     Available     true           2075762886ca222ec414398a9cdf72ad     2024-09-04T12:22:20Z     2024-09-03T12:20:20Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-productpage                  Root     Available     true           733ce926e70e263a0e03928d910176d7     2034-07-27T09:58:47Z     2024-07-29T09:58:47Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-ratings                      Leaf     Available     true           e67e3ba9bb12a54d728cd2b11e52e279     2024-09-04T12:22:20Z     2024-09-03T12:20:20Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-ratings                      Root     Available     true           733ce926e70e263a0e03928d910176d7     2034-07-27T09:58:47Z     2024-07-29T09:58:47Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-reviews                      Leaf     Available     true           fd635ede6043d719bb2230242b4e5eeb     2024-09-04T12:22:20Z     2024-09-03T12:20:20Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-reviews                      Root     Available     true           733ce926e70e263a0e03928d910176d7     2034-07-27T09:58:47Z     2024-07-29T09:58:47Z
```

Did I do something wrong? 
I set this in `meshConfig`:

```
meshConfig:
  discoverySelectors:
     - matchLabels:
            something/istio-enabled: "true"
```
This worked as a charm, I wanted Kiali to exclude non-istio namespaces and they recommend to use discoverySelectors. That worked as expected besides the CA doesn't work anymore.

**Comment by kovaxur** [comment]
As I remember, for me it fixed when I added the istio-system namespace to be selected by the discoverySelector

**Comment by dan-m8t** [comment]
I can confirm that fixed my issue, I labled `kube-system` with the label from `discoverySelector` and deleted the ztunnel pods again (not sure if this is neccessary?):

```
istioctl -i kube-system zc certificates ztunnel-tn4nl                                                                      
CERTIFICATE NAME                                                        TYPE     STATUS        VALID CERT     SERIAL NUMBER                        NOT AFTER                NOT BEFORE
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-details         Leaf     Available     true           f5c5e85d41a56a027cd8284694f3e52e     2024-09-04T12:36:19Z     2024-09-03T12:34:19Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-details         Root     Available     true           733ce926e70e263a0e03928d910176d7     2034-07-27T09:58:47Z     2024-07-29T09:58:47Z
spiffe://cluster.local/ns/istio-demo/sa/bookinfo-productpage     Leaf     Available     true           bf4ba78da9a109fb088af2f351f95fa8     2024-09-04T12:36:19Z     2024-09-03T12:34:19Z
```
Thanks!

**Comment by prithvixi** [comment]
> Solved it. I was deploying Istio on the default namespace instead of `istio-system`, I had to modify `istiod`'s values with:
> 
> ```yaml
> istio-istiod:
>   profile: ambient
>   pilot:
>     env:
>       CA_TRUSTED_NODE_ACCOUNTS: "default/ztunnel,kube-system/ztunnel"
> ```

Thanks it really worked.
